### PR TITLE
complete refactor of hii water task

### PR DESCRIPTION
**Pull request by thanrobinson**

### CODE REVIEW

**Comment by thanrobinson** [comment]
[hii_water temporal automation](https://trello.com/c/KxHtIWVn/271-hiiwater-temporal-automation)

File: src/hii_water.py [hunk 1]
```diff
@@ -5,24 +5,29 @@
 
 
 class HIIWater(HIITask):
-    ee_rootdir = "projects/HII/v1"
-    ee_driverdir = "driver/water"
-    # TODO: figure out inputs that will be updated
+    scale = 300
+    gpw_cadence = 5
+    wide_river_width = 30  # meters
+    coastal_settlements_dist = 4000  # meters
```
Comment (saanobhaai) [inline src/hii_water.py]: might as well spell out `coastal_settlements_distance`

File: src/hii_water.py [hunk 2]
```diff
@@ -5,24 +5,29 @@
...
+    indirect_distance = 15000  # meters
+    DECAY_CONSTANT = -0.0002
```
Comment (saanobhaai) [inline src/hii_water.py]: Doesn't really matter but generally the convention is "constants" (which these aren't really, in python, but they're marked as such by upper case) before variables.

File: src/hii_water.py [hunk 3]
```diff
@@ -35,139 +40,133 @@ class HIIWater(HIITask):
...
-        self.set_aoi_from_ee('projects/HII/v1/source/realms/' + self.realm)  
+        self.gpw = ee.ImageCollection(self.inputs["gpw"]["ee_path"])
+        self.jrc = ee.Image(self.inputs["jrc"]["ee_path"])
+        self.ocean = ee.Image(self.inputs["ocean"]["ee_path"])
+        self.caspian_sea_fc = ee.FeatureCollection(
+            self.inputs["caspian_sea"]["ee_path"]
+        )
+        self.watermask = ee.Image(self.inputs["watermask"]["ee_path"])
+        self.ee_taskdate = ee.Date(self.taskdate.strftime(self.DATE_FORMAT))
+        self.kernel = {
```
Comment (saanobhaai) [inline src/hii_water.py]: plural?

File: src/hii_water.py [hunk 4]
```diff
@@ -35,139 +40,133 @@ class HIIWater(HIITask):
...
+        elif end_test.getInfo():
+            gpw_taskdate = self.gpw_latest()
+        elif start_test.getInfo():
+            gpw_taskdate = self.gpw_earliest()
+        else:
+            raise Exception("no valid GPW image")
+
+        caspian_sea = self.caspian_sea_fc.reduceToImage(["glwd_id"], ee.Reducer.max())
+
+        ocean = ee.Image.cat(self.ocean.eq(0), caspian_sea.eq(1)).reduce(
```
Comment (saanobhaai) [inline src/hii_water.py:129]: Odd that it's where ocean == 0, but I guess that's just the nature of the data

File: src/hii_water.py [hunk 5]
```diff
@@ -35,139 +40,133 @@ class HIIWater(HIITask):
...
+        end_test = ee_taskdate_millis.gt(gpw_last_date)
+        interpolate_test = start_test.eq(0).And(end_test.eq(0))
+        if interpolate_test.getInfo():
+            gpw_taskdate = self.gpw_interpolated()
+        elif end_test.getInfo():
+            gpw_taskdate = self.gpw_latest()
+        elif start_test.getInfo():
+            gpw_taskdate = self.gpw_earliest()
+        else:
+            raise Exception("no valid GPW image")
```
Comment (saanobhaai) [inline src/hii_water.py:125]: I haven't looked, but I assume this is all the same as the GPW code we have elsewhere, right? If so we should make a card to refactor it out into `task_base`

File: src/hii_water.py [hunk 6]
```diff
@@ -5,24 +5,29 @@
...
+    scale = 300
+    gpw_cadence = 5
+    wide_river_width = 30  # meters
+    coastal_settlements_dist = 4000  # meters
+    indirect_distance = 15000  # meters
+    DECAY_CONSTANT = -0.0002
+    INDIRECT_INFLUENCE = 4
+
     inputs = {
         "jrc": {
```
Comment (saanobhaai) [inline src/hii_water.py]: I've never liked this name -- how about `surface_water` or `GSW`?

File: src/hii_water.py [hunk 7]
```diff
@@ -35,139 +40,133 @@ class HIIWater(HIITask):
...
-            reducer=ee.Reducer.max(),
-            kernel=ee.Kernel.circle(self.coast_settle_max / 2, "meters"),
-        ).reproject(crs=self.crs, scale=self.scale)
-
-        minpix_4km_threshold = (
-            within_min_of_inland_coast.selfMask()
-            .connectedPixelCount(MIN_PIX_COUNT)
-            .reproject(crs=self.crs, scale=THRESHOLD_SCALE)
+        jrc_water = (
+            self.jrc.select("occurrence").lte(70).unmask(1).updateMask(ocean.eq(0))
```
Comment (saanobhaai) [inline src/hii_water.py]: `70` should be a class property variable

**Review by saanobhaai — CHANGES_REQUESTED** [review]
Just a few little things... I didn't actually try to reason out all the masking and unmasking because my poor brain would pop, but the spatial results you undoubtedly have proven anyway.

File: src/hii_water.py (hunk 5 above)
Comment (thanrobinson) [reply src/hii_water.py:125]: I added a card to trello

File: src/hii_water.py (hunk 2 above)
Comment (saanobhaai) [reply src/hii_water.py]: Sorry, just meant to change the order, not capitalize these... `scale` is actually a variable property from `task_base` so you want to use it, lowercase.

File: src/hii_water.py [hunk 8]
```diff
@@ -5,16 +5,17 @@
...
-    indirect_distance = 15000  # meters
+    SCALE = 300
+    GPW_CADENCE = 5
+    WATER_OCCURANCE_THRESHOLD = 70
+    WIDE_RIVER_MIN_WIDTH = 30  # meters
+    SETTLMENT_DISTANCE_FROM_COAST = 4000  # meters
+    INDIRECT_DISTANCE = 15000  # meters
     DECAY_CONSTANT = -0.0002
-    INDIRECT_INFLUENCE = 4
+    INDIRECT_INFLUENCE = 4  # TODO: Set to 10.
```
Comment (saanobhaai) [inline src/hii_water.py]: Why not just do this?

File: src/hii_water.py [hunk 9]
```diff
@@ -5,16 +5,17 @@
...
-    scale = 300
-    gpw_cadence = 5
-    wide_river_width = 30  # meters
-    coastal_settlements_dist = 4000  # meters
-    indirect_distance = 15000  # meters
+    SCALE = 300
+    GPW_CADENCE = 5
+    WATER_OCCURANCE_THRESHOLD = 70
+    WIDE_RIVER_MIN_WIDTH = 30  # meters
+    SETTLMENT_DISTANCE_FROM_COAST = 4000  # meters
```
Comment (saanobhaai) [inline src/hii_water.py]: spelling

File: src/hii_water.py [hunk 10]
```diff
@@ -5,16 +5,17 @@
...
 
 class HIIWater(HIITask):
-    scale = 300
-    gpw_cadence = 5
-    wide_river_width = 30  # meters
-    coastal_settlements_dist = 4000  # meters
-    indirect_distance = 15000  # meters
+    SCALE = 300
+    GPW_CADENCE = 5
+    WATER_OCCURANCE_THRESHOLD = 70
```
Comment (saanobhaai) [inline src/hii_water.py]: spelling

**Review by saanobhaai — CHANGES_REQUESTED** [review]
Even littler things.

**Pull request merged**